### ISSUE TITLE
C# compiler should not report CS8780 when a union pattern declares a variable while matching a struct case through an open union type

### ISSUE THREAD

**Issue by solarizedalias** [dotnet/roslyn#84883]
## Version used

`.NET SDK 11.0.100-preview.7.26381.103` (C# 15 preview, `<LangVersion>preview</LangVersion>`)

## Summary

Since Preview 7, a union pattern that declares a variable is rejected with `CS8780` when the case type is a **struct**.
Declaring the very same case type as a **class** compiles.

The second condition is on the union, not the case: the union type must have an open type parameter.
In the repro below no case mentions that parameter, yet its presence alone is enough.
The same struct case matched through a closed union type compiles.

Preview 6 accepts all of the code below.

## Steps to reproduce

```csharp
public record class RefCase(int Value);

public record struct ValCase(int Value);

public union RefUnion<T>(RefCase);

public union ValUnion<T>(ValCase);

public static class Repro
{
    // compiles
    public static int FromRef<T>(RefUnion<T> u) => u switch { RefCase(var v) => v };

    // error CS8780 on `var v` -- identical to FromRef apart from class vs struct
    public static int FromVal<T>(ValUnion<T> u) => u switch { ValCase(var v) => v };

    // compiles -- same union and same pattern as FromVal, closed type argument
    public static int FromValClosed(ValUnion<int> u) => u switch { ValCase(var v) => v };
}
```

Build against `net11.0` with `<LangVersion>preview</LangVersion>`.
No other files, usings or project settings are needed.

## Expected behavior

All three methods compile, as they do on Preview 6.

Preview 7's Try-Both matching lets a union pattern match either the union instance itself or its underlying value.
When both arms are live the match is `or`-like, which is what `CS8780` forbids a variable designation in.

The [unions specification](https://github.com/dotnet/csharplang/blob/05eb4800fc3dc76259ccd49ac01f1eeb49222380/proposals/csharp-15.0/unions.md#type-pattern) enumerates four cases for a *type_pattern* against a union value, and only one of them is `or`-shaped:

> If the union type itself is *pattern compatible* with *type* and union `Value` is *pattern compatible* with *type*, *type_pattern* is equivalent to ```type or { Value: type }```, assuming that this form doesn't use any special treatment for unions.
>
> If the union type itself is not *pattern compatible* with *type* and union `Value` is *pattern compatible* with *type*, *type_pattern* is equivalent to ```{ Value: type }```, assuming that this form doesn't use any special treatment for unions.

`ValUnion<T>` and `ValCase` are unrelated struct types, so a `ValUnion<T>` can never itself be a `ValCase`.
That is the second case: the pattern is equivalent to `{ Value: ValCase }`, which contains no `or`.
There is no second arm to make the match `or`-like, so the designation should be permitted — exactly as it is in `FromRef` and `FromValClosed`.
(`ValCase(var v)` is a positional pattern, equivalent to ```ValCase and (var v)``` under the same document's [Positional pattern](https://github.com/dotnet/csharplang/blob/05eb4800fc3dc76259ccd49ac01f1eeb49222380/proposals/csharp-15.0/unions.md#positional-pattern) section, so its type portion is governed by the rule above.)

The specification conditions the `or`-shaped form on pattern compatibility between the union type and the case type.
It says nothing about whether the union type has an open type parameter, yet that is the only property separating the rejected method from the two accepted ones.

## Actual behavior

One error, on `FromVal` only:

> Repro.cs(15,75): error CS8780: A variable may not be declared within a 'not' or an 'or' pattern or a union matching involving matching against either the instance, or its underlying value.

`FromRef` and `FromValClosed` compile without diagnostics in the same compilation.

## Why this matters

Both references below are to [Union types](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/union).

**Section "Union declarations"** gives this as its example of a generic union:

```csharp
public record class None;
public record class Some<T>(T Value);
public union Option<T>(None, Some<T>);
```

That compiles.
Declaring `None` and `Some<T>` as `record struct` instead makes any generic method that binds a variable while matching an open `Option<T>` fail with `CS8780`, while the same methods over `Option<int>` keep compiling.

**Section "Custom union types", subsection "Non-boxing access pattern"** describes `HasValue` plus a `TryGetValue` per case, and states:

> The compiler prefers `TryGetValue` over the `Value` property when implementing pattern matching, which avoids boxing value types.

That facility exists specifically so value-type cases can be matched without boxing, so a union that adopts it for that purpose has struct case types — exactly the shape this rejects.

Calling `TryGetValue` directly is a workaround, at the cost of the pattern syntax.

## Additional data points

Each row was built separately, changing one thing at a time from the repro above:

| variant | result |
|---|---|
| struct case inside a **non-generic** union | compiles |
| custom `[Union]` type instead of the `union X(...)` declaration | **CS8780** |
| with or without the non-boxing `HasValue` / `TryGetValue` members | **CS8780** |
| type parameter from the containing type rather than the method | **CS8780** |

The first row complements `FromValClosed`: a struct case compiles inside a union that is not generic at all, as well as inside a closed generic one. Only an open union rejects it.

How the union is declared makes no difference.

Dropping `Value` to remove the value arm is not available — it is required (`CS9386`).

## Regression range

| SDK | result |
|---|---|
| `11.0.100-preview.6.26359.118` | compiles |
| `11.0.100-preview.7.26381.103` | **CS8780** |

Preview 7 is where Try-Both merged (#84531, plus #84323 / #84365 / #84418 for type, var/declaration and recursive patterns).
It is not mentioned in the Preview 7 release notes, and reference-type cases are not restricted, which is why this looks like a side effect rather than an intended restriction.

## Possible cause

The check for whether the instance arm of Try-Both is viable appears not to be discharged whenever the union type has an open type parameter — even when, as in `FromVal`, the matched case type is not generic and the parameter is unused — leaving both arms live and so treating the match as `or`-like.
For `ValUnion<int>` the arm is discharged and the designation is allowed.

**Comment by AlekseyTs** [comment]
The mechanism used to detect whether input is *pattern compatible* with a type is imprecise in presence of open generic types. This can lead to Try-Both matching applied in such cases. The error is a known limitation of the current implementation of Try-Both matching scenario. Due to other reasons we decided to remove Try-Both matching for unions. Once that happens, the error won't be reported.
